src/data/data.cc

```diff
@@ -829,48 +830,68 @@ uint64_t SparsePage::Push(const AdapterBatchT& batch, float missing, int nthread
           last_line.GetElement(last_line.Size() - 1).row_idx - base_rowid;
     }
   }
-  builder.InitBudget(expected_rows, nthread);
+  size_t batch_size = batch.Size();
+  const size_t thread_size = batch_size/nthread;
+  builder.InitBudget(expected_rows+1, nthread);
   uint64_t max_columns = 0;
+  if (batch_size == 0) {
+    omp_set_num_threads(nthread_original);
+    return max_columns;
+  }
+  std::vector<std::vector<uint64_t>> max_columns_vector(nthread);
 
   // First-pass over the batch counting valid elements
-  size_t batch_size = batch.Size();
-#pragma omp parallel for schedule(static)
-  for (omp_ulong i = 0; i < static_cast<omp_ulong>(batch_size);
-       ++i) {  // NOLINT(*)
+#pragma omp parallel num_threads(nthread)
+  {
     int tid = omp_get_thread_num();
-    auto line = batch.GetLine(i);
-    for (auto j = 0ull; j < line.Size(); j++) {
-      data::COOTuple element = line.GetElement(j);
-      max_columns =
-          std::max(max_columns, static_cast<uint64_t>(element.column_idx + 1));
-      if (!common::CheckNAN(element.value) && element.value != missing) {
-        size_t key = element.row_idx - base_rowid;
-        // Adapter row index is absolute, here we want it relative to
-        // current page
-        CHECK_GE(key, builder_base_row_offset);
-        builder.AddBudget(key, tid);
+    size_t begin = tid*thread_size;
+    size_t end = tid != (nthread-1) ? (tid+1)*thread_size : batch_size;
+    max_columns_vector[tid].resize(1, 0);
+    uint64_t& max_columns_local = max_columns_vector[tid][0];
+
+    for (size_t i = begin; i < end; ++i) {
+      auto line = batch.GetLine(i);
+      for (auto j = 0ull; j < line.Size(); j++) {
+        auto element = line.GetElement(j);
+        const size_t key = element.row_idx - base_rowid;
+        CHECK_GE(key,  builder_base_row_offset);
+        max_columns_local =
+            std::max(max_columns_local, static_cast<uint64_t>(element.column_idx + 1));
+
+        if (!common::CheckNAN(element.value) && element.value != missing) {
+          // Adapter row index is absolute, here we want it relative to
+          // current page
+          builder.AddBudget(key, tid);
+        }
       }
     }
   }
+  for (const auto & max : max_columns_vector) {
+    max_columns = std::max(max_columns, max[0]);
+  }
+
   builder.InitStorage();
 
   // Second pass over batch, placing elements in correct position
-#pragma omp parallel for schedule(static)
-  for (omp_ulong i = 0; i < static_cast<omp_ulong>(batch_size);
-       ++i) {  // NOLINT(*)
+
+#pragma omp parallel num_threads(nthread)
+  {
```

Comment:
Please help adding `dmlc::OMPException` to capture exception happened inside omp block.

Reply:
applied